mapper/pchain/tx_parser.go

```
@@ -0,0 +1,740 @@
+		inputTxAccounts = make(map[string]*types.AccountIdentifier)
+	}
+
+	return &TxParser{
+		isConstruction:  isConstruction,
+		hrp:             hrp,
+		chainIDs:        chainIDs,
+		inputTxAccounts: inputTxAccounts,
+		dependencyTxs:   dependencyTxs,
+	}
+}
+
+func (t *TxParser) Parse(txID ids.ID, tx txs.UnsignedTx) (*types.Transaction, error) {
+	var ops *txOps
+	var txType string
+	var err error
+	switch unsignedTx := tx.(type) {
+	case *txs.ExportTx:
+		txType = OpExportAvax
+		ops, err = t.parseExportTx(txID, unsignedTx)
+	case *txs.ImportTx:
+		txType = OpImportAvax
+		ops, err = t.parseImportTx(txID, unsignedTx)
+	case *txs.AddValidatorTx:
+		txType = OpAddValidator
+		ops, err = t.parseAddValidatorTx(txID, unsignedTx)
+	case *txs.AddDelegatorTx:
+		txType = OpAddDelegator
+		ops, err = t.parseAddDelegatorTx(txID, unsignedTx)
+	case *txs.RewardValidatorTx:
+		txType = OpRewardValidator
+		ops, err = t.parseRewardValidatorTx(unsignedTx)
+	case *txs.CreateSubnetTx:
+		txType = OpCreateSubnet
+		ops, err = t.parseCreateSubnetTx(txID, unsignedTx)
+	case *txs.CreateChainTx:
+		txType = OpCreateChain
+		ops, err = t.parseCreateChainTx(txID, unsignedTx)
+	case *txs.AddSubnetValidatorTx:
+		txType = OpAddSubnetValidator
+		ops, err = t.parseAddSubnetValidatorTx(txID, unsignedTx)
+	case *txs.AddPermissionlessValidatorTx:
+		txType = OpAddPermissionlessValidator
+		ops, err = t.parseAddPermissionlessValidatorTx(txID, unsignedTx)
+	case *txs.AddPermissionlessDelegatorTx:
+		txType = OpAddPermissionlessDelegator
+		ops, err = t.parseAddPermissionlessDelegatorTx(txID, unsignedTx)
+	case *txs.RemoveSubnetValidatorTx:
+		txType = OpRemoveSubnetValidator
+		ops, err = t.parseRemoveSubnetValidatorTx(txID, unsignedTx)
+	case *txs.TransformSubnetTx:
+		txType = OpTransformSubnetValidator
+		ops, err = t.parseTransformSubnetTx(txID, unsignedTx)
+	case *txs.AdvanceTimeTx:
+		txType = OpAdvanceTime
+		// no op tx
+	default:
+		log.Printf("unknown type %T", unsignedTx)
```

Should we return an error here?

---

No, we don't necessarily want to block parsing if we encounter new transaction types that's not supported yet in Rosetta.